src/crimson/os/seastore/segment_cleaner.h

```diff
@@ -887,7 +926,17 @@ class SegmentCleaner : public SegmentProvider {
       [this] {
 	blocked_io_wake = seastar::promise<>();
 	return blocked_io_wake->get_future();
-      });
+      }
+    ).then([this, projected_usage] {
+      ceph_assert(!blocked_io_wake);
+      projected_used_bytes += projected_usage;
```

Comment:
It looks to me that the following scenario is still possible:

1. transaction A and B are both waiting on SegmentCleaner::blocked_io_wake
2. segment cleaner set SegmentCleaner::blocked_io_wake's value
3. transaction A is scheduled and set projected_used_bytes, which leads to another round of blocked io wait
4. transaction B is scheduled and the assert here is triggered

Am I right?

If so, I think this [condition variable](https://github.com/ceph/ceph/pull/41803/commits/3207f6413c15edc91e46c9cd154a84c93c78bbc8) should be able to help:-)

Reply:
Ah, sorry, this is protected by write_pipeline.reserve_projected_usage, right?

Reply:
Right.